ldm/dream/args.py

```diff
@@ -0,0 +1,485 @@
+"""Helper class for dealing with image generation arguments.
+
+The Args class parses both the command line (shell) arguments, as well as the
+command string passed at the dream> prompt. It serves as the definitive repository
+of all the arguments used by Generate and their default values.
+
+To use:
+  a = Args()
+
+  # read in the command line options:
+  args = a.parse_args()
+
+  # read in a command passed to the dream> prompt:
+  opts = a.parse_cmd('do androids dream of electric sheep? -H256 -W1024 -n4')
+
+  # The args object acts like a namespace object
+  print(a.model)
+
+You can set attributes in the usual way, use vars(), etc.:
+
+  a.model = 'something-else'
+  do_something(**vars(a))
+
+It is helpful in saving metadata:
+
+  # To get a json representation of all the values, allowing
+  # you to override any values dynamically
+  j = a.json(seed=42)
+
+  # To get the prompt string with the switches, allowing you
+  # to override any values dynamically
+  j = a.prompt_str(seed=42)
+
+If you want to access the namespace objects from the shell args or the
+parsed command directly, you may use the values returned from the
+original calls to parse_args() and parse_cmd(), or get them later
+using the _arg_switches and _cmd_switches attributes. This can be
+useful if both the args and the command contain the same attribute and
+you wish to apply logic as to which one to use. For example:
+
+  a = Args()
+  args    = a.parse_args()
+  opts    = a.parse_cmd(string)
+  do_grid = args.grid or opts.grid
+
+To add new attributes, edit the _create_command_arg_parser() and
+_create_cmd_parser() methods.
+
+We also export the function build_metadata
+"""
+
+import argparse
+import shlex
+import json
+
+SAMPLER_CHOICES = [
+    'ddim',
+    'k_dpm_2_a',
+    'k_dpm_2',
+    'k_euler_a',
+    'k_euler',
+    'k_heun',
+    'k_lms',
+    'plms',
+]
+
+# is there a way to pick this up during git commits?
```

Comment:
I'd try to answer, but I'm not certain as to exactly what you're asking.

Reply:
> I'd try to answer, but I'm not certain as to exactly what you're asking.

There are two string constants under line 67:
~~~
APP_ID      = 'lstein/stable-diffusion'
APP_VERSION = 'v1.15'
~~~
These seem like things that git could do string substitution for us during the commit or release process. I used to be able to do something like this in Svn with $Id$ and $Author$.  Just a passing thought. Only the version will change frequently enough for it to be a chore.

Reply:
I *think* that this should be possible with a pre-commit hook (searching one out would be easier than writing one from scratch), but this has the same proeblem presented by *any* pre-commit hook - they're impossible to distribute and enforce along with a `git clone` operation.

i.e. we'd have to provide install and setup instructions for contributors... and hope that they follow them correctly - again, pre-commit hooks are **completely** client-side, so there's no auto-enforcement e.g. in this instance, a human would have to verify that those strings were replaced and replaced correctly.